e2e_playwright/st_dataframe_selections_test.py

```diff
@@ -0,0 +1,310 @@
+#     # submit the form
+#     app.get_by_test_id("stFormSubmitButton").first.locator("button").click()
+#     wait_for_app_run(app)
+#     # expected =  "Dataframe-in-form selection in session state: {'select': {'rows': [0, 2], 'columns': ['col_1', 'col_3', 'col_4']}}"
+#     expected = "Dataframe-in-form selection in session state:"
+#     selection_text = app.get_by_test_id("stMarkdownContainer").filter(has_text=expected)
+#     # wait_until(
+#     #     app,
+#     #     lambda: app.get_by_test_id("stMarkdownContainer")
+#     #     .filter(has_text=expected)
+#     #     .count()
+#     #     == 1,
+#     # )
+
+#     expect(selection_text).to_have_count(1)
+#     expect(app.get_by_test_id("stMarkdownContainer").nth(13)).to_have_text(
+#         "Dataframe-in-form selection in session state: {'select': {'rows': [0, 2], 'columns': ['col_1', 'col_3', 'col_4']}}",
+#         timeout=2000,
+#     )
+
+
+def test_in_form_selection_and_session_state(app: Page):
+    canvas = _get_in_form_df(app)
+    _select_some_rows_and_columns(app, canvas)
+
+    # nothing should be shown yet because we did not submit the form
+    expected = "Dataframe-in-form selection: {'select': {'rows': [], 'columns': []}}"
+    selection_text = app.get_by_test_id("stMarkdownContainer").filter(has_text=expected)
+    expect(selection_text).to_have_count(1)
+
+    # submit the form. Without the timeout the test is flaky, presumably because the submit-button does not have an eventlistener on time :/
+    app.wait_for_timeout(100)
+    app.get_by_test_id("baseButton-secondaryFormSubmit").click()
+    wait_for_app_run(app)
+
+    expected = "Dataframe-in-form selection: {'select': {'rows': [0, 2], 'columns': ['col_1', 'col_3', 'col_4']}}"
+    selection_text = app.get_by_test_id("stMarkdownContainer").filter(has_text=expected)
+    expect(selection_text).to_have_count(1)
+
+    expected = "Dataframe-in-form selection in session state: {'select': {'rows': [0, 2], 'columns': ['col_1', 'col_3', 'col_4']}}"
+    selection_text = app.get_by_test_id("stMarkdownContainer").filter(has_text=expected)
+    expect(selection_text).to_have_count(1)
+
+
+def test_multi_row_and_multi_column_selection_with_callback(app: Page):
+    canvas = _get_callback_df(app)
+    _select_some_rows_and_columns(app, canvas)
+
+    expected = "Dataframe selection callback: {'select': {'rows': [0, 2], 'columns': ['col_1', 'col_3', 'col_4']}}"
+    selection_text = app.get_by_test_id("stMarkdownContainer").filter(has_text=expected)
+    expect(selection_text).to_have_count(1)
+
+
+def test_multi_row_and_multi_column_select_snapshot(
+    app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Take a snapshot of multi-select to ensure visual consistency."""
+    canvas = _get_multi_row_and_column_select_df(app)
+    _select_some_rows_and_columns(app, canvas)
+    _expect_multi_row_multi_column_selection(app)
+
+    assert_snapshot(canvas, name="st_data_editor-multi_row_multi_column_selection")
```

Comment:
@raethlein 
nit: this needs to be `st_dataframe-` since the data editor does not support selections at the moment

Reply:
💡 gotcha